### ISSUE TITLE
Continuable Traps

### ISSUE THREAD

**Issue by ruipin** [RPCS3/rpcs3#5700]
Some games, such as most of the Naughty Dogs games, will contain various assertions that rely on PS3 timings or features that would be impossible to emulate correctly.

For example, TLOU will randomly fail with the following error in TTY:

```
******************************************************************************
*** ASSERTION: false && "Animation SPU jobs did not complete in time!"
***
***  Function: 
***      File: ndlib/anim/anim-mgr.cpp
***      Line: 1762
***     Built: 
******************************************************************************
```

This assertion triggers a "Trap!" exception in RPCS3 and causes the game to quit.

However, looking at the code which triggers the trap (0x00963028 for unpatched BCES version), the code is continuable - i.e., if we ignore the trap the game would work fine.

A patch to get rid of the above trap is:

    PPU-<hash>:
        - [ be32, 0x00963028, 0x60000000 ] # nop trap

With this patch, the above failure is still printed to TTY, but no actual crash occurs and the game works as expected.

Other games such as UC1 have very similar behaviour (in the case of UC1, it asserts against the RSX timestamp).

A lot of these behaviours are impossible to actually emulate correctly. We cannot guarantee that SPU jobs complete in a fixed time even with a supercomputer, since for example OS background tasks could kick in at any moment and influence SPU scheduling enough to cause crashes. Things like the RSX timestamp accuracy are also not possible to implement exactly like the PS3, due to various hardware limitations.

It is obviously possible to write game-specific patches for all of these issues, but that would mean a lot of patches and significant work per game. Since traps tend to be used for more harm than good, I would propose that we implement continuable traps:

1. Traps should not kill the game. They should simply pause the emulator, such that the user might resume the game if they want to.
2. There should be a configuration that downgrades traps to errors, meaning they don't pause the game by default.

**Comment by elad335** [comment]
Crysis 3 suffers froma similar issue as TLoU, see #3687. its random and purely related to timing.

**Comment by elad335** [comment]
Fixed by #7233

**Comment by VelocityRa** [comment]
@elad335 But #7233 degrades performance. Continuable traps would not, so perhaps this is still worth doing right?

**Comment by elad335** [comment]
Continuable traps is a horrible hack, imagine "continuable nullptr accesses".

**Comment by VelocityRa** [comment]
Realistically won't it be fine for the vast majority of games? Why else do games trap if not these timing checks?

I think it should be up to the user. I'm not saying it should be forced or even enabled by default.

So, I was thinking of only doing
```
2. There should be a configuration that downgrades traps to errors, meaning they
don't pause the game by default.
```
from the OP.

**Comment by elad335** [comment]
I think what really needs to be done is to implement ppu llvm executable code changes with patches, this would allow users to "nop-out" those traps if its really useful for specfic games even for llvm.
Patches would work with other kinds of assertations as well, kinds which this hack won't cover anyways.
> Why else do games trap in practice, except these timing checks?

Lots of reasons.. it's an assertation, could be triggered by emulation bugs etc.

**Comment by elad335** [comment]
Implemented by #8230, close issue.
